### PR TITLE
terraform: refactor ProvidedBy() to return an addrs.ProviderConfig interface

### PR DESCRIPTION
**Pull request by mildwonkey**
This refactor allows terraform to indicate whether a specific provider
configuration was found for the resource or if it is instead returning
the assumed default.

With that additional information the provider transformer can check if
there is a specific (non-default) provider FQN.

----

Another option I considered is having `ProvidedBy()` to return `nil` instead of an `AbsProviderConfig` when a specific ProviderConfig was not found (instead of return an assumed default provider) and making the provider transformer responsible for determining the provider in that case. I think the biggest argument in favor of this change is readability; I find the current behavior non-intuitive. It would require adding a DefaultProvider() function to the `GraphNodeProviderConsumer`.

### CODE REVIEW

**Comment by mildwonkey** [comment]
update to my earlier comment: I still like the idea of refactoring `ProvidedBy()` to return nil and add a `DefaultProvider()` function to `GraphNodeProviderConsumer` and plan to make a followup PR.

File: terraform/transform_attach_schema.go [hunk 1]
```diff
@@ -59,8 +60,16 @@ func (t *AttachSchemaTransformer) Transform(g *Graph) error {
 			mode := addr.Resource.Mode
 			typeName := addr.Resource.Type
 			providerAddr, _ := tv.ProvidedBy()
+			var providerFqn addrs.Provider
 
-			schema, version := t.Schemas.ResourceTypeConfig(providerAddr.Provider, mode, typeName)
+			switch providerAddr.(type) {
```
Comment (jbardin) [inline terraform/transform_attach_schema.go]: you can assign a value in a type switch, e.g.

`switch p := providerAddr.(type) {`

Then use it in each case with the correct type:

`providerFqn = p.Provider`

Comment (mildwonkey) [reply terraform/transform_attach_schema.go]: oh hey thanks! That's so much more readable!

**Comment by mildwonkey** [comment]
Whoops! sorry for the confusion @apparentlymart and tyvm

**Pull request merged**